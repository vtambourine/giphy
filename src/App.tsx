import { throttle } from 'lodash';
import cx from 'classnames';
import React from 'react';

import Feed from './Feed';
import Provider from './provider';
import Search from './Search';
import Spinner from './Spinner';

import styles from './App.module.css';

interface IAppProps {}

interface IAppState {
  isLoading: boolean;
  data: GIF[];
}

class App extends React.Component<IAppProps, IAppState> {
  state = {
    isLoading: true,
    data: [] as GIF[],
  };

  throttledSubmit: (query: string) => void;
  provider?: Provider;

  constructor(props: IAppProps) {
    super(props);
    this.throttledSubmit = throttle(this.onSubmit, 1500, { leading: false });
  }

  onSubmit = (query: string) => {
    this.setState({
      isLoading: true,
    });
    this.provider = new Provider(query);
    this.provider.next().then(({ data }) =>
      this.setState({
        isLoading: false,
        data,
      }),
    );
  };

  onFeedEnd = () => {
    if (!this.state.isLoading && this.provider) {
      this.setState({
        isLoading: true,
      });

      this.provider.next().then(({ data }) =>
        this.setState({
          isLoading: false,
          data: [...this.state.data, ...data],
        }),
      );
    }
  };

  render() {
    const { data, isLoading } = this.state;
    return (
      <div className={styles.app}>
        <div
          className={cx(styles.container, {
            [styles.empty]: data.length === 0,
            [styles.loading]: isLoading,
          })}
        >
          <div className={styles.search}>
            <Search onChange={this.throttledSubmit} />
            {isLoading && (
              <div className={styles.spinner}>
                <Spinner />
              </div>
            )}
          </div>
          <div className={styles.feed}>
            <Feed data={data} onFeedEnd={this.onFeedEnd} loading={isLoading} />
          </div>
        </div>
      </div>
    );
  }
}

export default App;
