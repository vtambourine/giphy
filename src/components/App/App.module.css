:root {
  --search-height: 60px;
}

.app {
  display: flex;
  justify-content: center;
  background-color: #121212;
}

.container {
  display: flex;
  flex-direction: column;
  width: 480px;
  height: 100vh;
  overflow: hidden;
}

.search {
  position: relative;
  height: var(--search-height);
  margin: 12px 0;
  transition: transform 0.3s;
}

.feed {
  flex: 1;
  height: calc(100vh - var(--search-height));
}

.empty .search {
  flex: 1;
  height: 100vh;
  transform: translateY(50%);
}

.empty .feed {
  height: 0;
  flex: 0 1;
}

.loading .feed {
  opacity: 0.3;
}

.spinner {
  position: absolute;
  left: 0;
  top: 0;
  margin-left: 50%;
  transform: translateX(-50%) translateY(150%);
}
