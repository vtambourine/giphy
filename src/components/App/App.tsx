import { throttle } from 'lodash';
import cx from 'classnames';
import React from 'react';

import { Feed, Search, Spinner } from '..';
import { Provider } from '../../services';

import styles from './App.module.css';

interface IAppProps {}

interface IAppState {
  loadingFeed: boolean;
  loadingQuery: boolean;
  data: GIF[];
}

class App extends React.Component<IAppProps, IAppState> {
  state = {
    loadingFeed: false,
    loadingQuery: false,
    data: [] as GIF[],
  };

  throttledSubmit: (query: string) => void;
  provider?: Provider;

  constructor(props: IAppProps) {
    super(props);
    this.throttledSubmit = throttle(this.onSubmit, 1500, { leading: false });
  }

  onSubmit = (query: string) => {
    this.setState({
      loadingQuery: true,
    });
    this.provider = new Provider(query);
    this.provider.next().then(({ data }) =>
      this.setState({
        loadingQuery: false,
        data,
      }),
    );
  };

  onFeedEnd = () => {
    if (!this.state.loadingQuery && this.provider) {
      this.setState({
        loadingFeed: true,
      });
      this.provider.next().then(({ data }) =>
        this.setState({
          loadingFeed: false,
          data: [...this.state.data, ...data],
        }),
      );
    }
  };

  render() {
    const { data, loadingFeed, loadingQuery } = this.state;
    return (
      <div className={styles.app}>
        <div
          className={cx(styles.container, {
            [styles.empty]: data.length === 0,
            [styles.loading]: loadingQuery,
          })}
        >
          <div className={styles.search}>
            <Search onChange={this.throttledSubmit} />
            {loadingQuery && (
              <div className={styles.spinner}>
                <Spinner />
              </div>
            )}
          </div>
          <div className={styles.feed}>
            <Feed
              data={data}
              onFeedEnd={this.onFeedEnd}
              loading={loadingFeed}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default App;
