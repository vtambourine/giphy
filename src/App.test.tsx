import React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import App from './App';

describe('App', () => {
  const mockData = [
    {
      title: 'Title',
      images: {
        fixed_width_still: {
          width: '200',
          url: 'http://url.com',
        },
      },
    },
  ];

  let wrapper: ReactWrapper<any, any, App>;

  beforeEach(() => {
    wrapper = mount(<App />);
  });

  it('renders correctly', () => {
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('renders loading indicator', () => {
    wrapper.setState({
      loadingQuery: true,
    });

    expect(wrapper.find('Spinner')).toHaveLength(1);
  });

  it('renders images feed', () => {
    wrapper.setState({
      data: mockData,
    });

    expect(wrapper.find('Spinner')).toHaveLength(0);
  });
});
