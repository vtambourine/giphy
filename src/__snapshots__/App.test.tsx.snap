// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "app",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "container empty",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "search",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "search",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "input",
                      "placeholder": "Search GIPHY...",
                      "type": "text",
                      "value": "",
                    },
                    "children": Array [],
                    "name": "input",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                      "placeholder": undefined,
                      "type": undefined,
                      "value": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                      "placeholder": undefined,
                      "type": undefined,
                      "value": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "feed",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "feed",
                  },
                  "children": Array [],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "feed",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "feed",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "search",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "search",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "input",
                        "placeholder": "Search GIPHY...",
                        "type": "text",
                        "value": "",
                      },
                      "children": Array [],
                      "name": "input",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                        "placeholder": undefined,
                        "type": undefined,
                        "value": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                        "placeholder": undefined,
                        "type": undefined,
                        "value": undefined,
                      },
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
